shared/src/service_registry.rs

```diff
@@ -0,0 +1,215 @@
+        gas_price_estimation::FakeGasPriceEstimator,
+        price_estimate::{BaselinePriceEstimator, PriceEstimating},
+        sources::uniswap::pool_fetching::{MockPoolFetching, PoolFetching},
+    };
+    use ethcontract::H160;
+    use gas_estimation::GasPriceEstimating;
+    use model::order::OrderKind;
+    use std::collections::HashSet;
+
+    trait Fooable: Send + Sync {
+        fn foo(&self) -> i32;
+    }
+
+    struct Foo;
+    impl Fooable for Foo {
+        fn foo(&self) -> i32 {
+            42
+        }
+    }
+    impl ServiceInitializable for Foo {
+        fn init(_: &ServiceRegistry) -> Self {
+            Foo
+        }
+    }
+    impl_service_linkable!(Foo => Fooable);
+
+    struct Bar(Arc<dyn Fooable>);
+    impl Bar {
+        fn magic(&self) -> i32 {
+            (self.0.foo() * 191) / 6
+        }
+    }
+    impl ServiceInitializable for Bar {
+        fn init(registry: &ServiceRegistry) -> Self {
+            Bar(registry.get())
+        }
+    }
+
+    #[test]
+    fn can_register_and_get() {
+        let mut registry = ServiceRegistry::default();
+        registry
+            // Register a `Foo` implementation for a `Fooable` component.
+            .register_as::<Foo, dyn Fooable>()
+            // Register a concrete `Bar` component.
+            .register::<Bar>();
+
+        assert_eq!(registry.get::<dyn Fooable>().foo(), 42);
+        assert_eq!(registry.get::<Bar>().magic(), 1337);
+    }
+
+    impl_service_linkable!(MockPoolFetching => PoolFetching);
+    impl_service_linkable!(FakeGasPriceEstimator => GasPriceEstimating);
+    impl_service_linkable!(MockBadTokenDetecting => BadTokenDetecting);
+    impl_service_linkable!(BaselinePriceEstimator => PriceEstimating);
+
+    // This would represent the command line arguments.
+    struct Options {
+        base_tokens: HashSet<H160>,
+        native_token: H160,
+    }
+
+    impl ServiceInitializable for BaselinePriceEstimator {
+        fn init(registry: &ServiceRegistry) -> Self {
+            let options = registry.get::<Options>();
+            Self::new(
+                registry.get(),
+                registry.get(),
+                options.base_tokens.clone(),
+                registry.get(),
+                options.native_token,
+            )
+        }
+    }
+
+    #[tokio::test]
+    async fn register_balancer_component_with_mocks() {
+        let mut token_detector = MockBadTokenDetecting::new();
+        token_detector
+            .expect_detect()
+            .returning(|_| Ok(TokenQuality::Good));
+
+        let mut registry = ServiceRegistry::default();
+        registry
+            .add(Options {
+                base_tokens: Default::default(),
+                native_token: Default::default(),
+            })
+            .add(token_detector)
+            .link::<MockBadTokenDetecting, dyn BadTokenDetecting>()
+            .register_as::<MockPoolFetching, dyn PoolFetching>()
```

Comment:
I'm not entirely clear on the differences between the three things here. Is register as essentially a combination of add and link?

Reply:
Yes, and register requires the type to implement `ServiceInitializable` so that it can create the instance on its behalf.